Solve ode (ordinary differential equation) guide page (#23916)
sympy/sympy: doc/src/guides/solving/solve-ode.md

# Solve an Ordinary Differential Equation Algebraically

Use SymPy to solve an ordinary differential equation algebraically. For example,
solving $y''(x) + 9y(x)=0 $ yields $ y(x)=C_{1} \sin(3x)+ C_{2} \cos(3x)$.

Alternatives to consider:
- *which SciPy functions? from

>   https://docs.scipy.org/doc/scipy/reference/integrate.html?highlight=ode?*

Reviewer: I think solve_ivp is good for most instances https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.solve_ivp.html#scipy.integrate.solve_ivp. Or we can just reference this page https://docs.scipy.org/doc/scipy/reference/integrate.html#solving-initial-value-problems-for-ode-systems

result, which you can then consider. The simplest way to disable integration is
with the `all_Integral` hint because you do not need to know which hint to
supply: for any hint with a corresponding `_Integral` hint, `all_Integral` only
returns the `_Integral` hint.

### Select a Specific Solver

You may wish to select a specific solver using a hint for a couple of reasons:
- educational purposes: for example if you are learning about a specific method
  to solve ODEs and want to get a result that exactly matches that method
- form of the result: sometimes an ODE can be solved by many different solvers,
  and they can return different results. They will be mathematically equivalent,
  though the arbitrary constants may not be. {func}`~.dsolve` by default tries
  to use the "best" solvers first, which are most likely to return the most
  usable output, but it is not a perfect heuristic. For example, the "best"
  solver may produce a result with an integral that SymPy cannot solve, but
  another solver may produce a different integral that SymPy can solve. So if
  the solution isn't in a form you like, you can try other hints to check

>   whether they give a preferable result.

Author: > Good examples of this [using different solvers that can be applied to the same ODE] are ODEs that match the homogeneous_order hints

@asmeurer is it worth giving an example of this? If so, is one of the homogeneous order equations given in the [`homogeneous_order`](https://docs.sympy.org/dev/modules/solvers/ode.html?highlight=ode#homogeneous-order) docstring a good example, either `f(x)` or `x**2*f(x)/sqrt(x**2+f(x)**2)`?

Author: @asmeurer, reminding you to please construct an example where the form of the ODE solution depends on the hint given.

## Not All Equations Can Be Solved

### Equations With No Solution

*Equations with no solution content*

> ### Equations With No Analytical Solution

Reviewer: This looks like a bug, although for a system you should specify the list of functions as the second argument to dsolve (also a good idea for single ODEs, although it will do it automatically if there is only one function in the expression).

> ### Equations With No Analytical Solution

Reviewer 2: That's one of the legacy solvers that are mostly just broken and should probably just be removed:
https://github.com/sympy/sympy/wiki/ODE-Systems-roadmap#nonlinear-solvers---eliminating-the-independent-variable

> ### Equations With No Analytical Solution

Author: Thanks; I added that you supply a list of functions to solve for.

I'm still seeking an example of an ODE (or system of) where there isn't a closed-form solution (and SymPy recognizes that), and would appreciate any ideas!

> ### Equations With No Analytical Solution

Reviewer: Maybe a double pendulum. See https://scienceworld.wolfram.com/physics/DoublePendulum.html (eqs 14 and 19). Here I simplified them by setting the masses and lengths to 1, although I'm sure it would give the same thing too if I included them:

```py
>>> theta1, theta2 = symbols('theta1 theta2', cls=Function)
>>> g = symbols('g')
>>> eq1 = 2*theta1(t).diff(t, t) + theta2(t).diff(t, t)*cos(theta1(t) - theta2(t)) + theta2(t).diff(t)**2*sin(theta1(t) - theta2(t)) + 2*g*sin(theta1(t))
>>> eq2 = theta2(t).diff(t, t) + theta1(t).diff(t, t)*cos(theta1(t) - theta2(t)) - theta1(t).diff(t)**2*sin(theta1(t) - theta2(t)) + g*sin(theta2(t))
>>> dsolve([eq1, eq2], [theta1(t), theta2(t)])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/./sympy/solvers/ode/ode.py", line 589, in dsolve
    raise NotImplementedError
NotImplementedError
```

(please double check that I transcribed it correctly)

*Equations with no analytical solution content*

### Equations Which Have An Analytical Solution, and SymPy Cannot Solve

> *Equations which have an analytical solution, and SymPy cannot solve content*

Author: Can someone (perhaps @oscarbenjamin) come up with an ODE (one or system) which has a closed-form solution, but which SymPy cannot solve?

Reviewer 2: Any example like that will always be fragile because in principle a future version of SymPy will be able to solve it. In fact putting it in the guide makes it particularly likely that `dsolve` would be improved in future to handle that case.

there may be an opportunity to simplify the constants:

```py
>>> from sympy import Function, dsolve, Derivative, symbols
>>> from sympy.abc import x
>>> from sympy.solvers.ode.ode import constantsimp
>>> y = Function('y')
>>> result = dsolve(Derivative(y(x), x, x) + 9*y(x), y(x)); result
Eq(y(x), C1*sin(3*x) + C2*cos(3*x))
>>> sol = result.rhs; sol
C1*sin(3*x) + C2*cos(3*x)
>>> ex = sol + (3 * sol / 2); ex
5*C1*sin(3*x)/2 + 5*C2*cos(3*x)/2
>>> C1, C2, C3 = symbols("C1 C2 C3")
>>> constantsimp(ex, {C1, C2, C3})
C1*sin(3*x) + C2*cos(3*x)
```

After simplifying the arbitrary constants, those constants may be
non-consecutive ordered. For example the simplified result may contain `C1` and
`C3` but not `C2`. You can fix this using {func}`~.constant_renumber`:

```py
>>> from sympy import symbols, cos, sin
>>> from sympy.solvers.ode.ode import constantsimp, constant_renumber
>>> C1, C2, C3 = symbols("C1 C2 C3")
>>> expr = C1*cos(x) + C2*cos(x) + C3*sin(x); expr
C1*cos(x) + C2*cos(x) + C3*sin(x)
>>> simplified = constantsimp(expr, {C1, C2, C3}); simplified
C1*cos(x) + C3*sin(x)
>>> constant_renumber(simplified)
C1*sin(x) + C2*cos(x)

> ```

Author: I was trying to come up with realistic examples of when you would work with arbitrary constants. Perhaps it's my ignorance, but I didn't find the examples in [`constantsimp`](https://docs.sympy.org/dev/modules/solvers/ode.html?highlight=constantsimp#sympy.solvers.ode.constantsimp) or [`constant_renumber`](https://docs.sympy.org/dev/modules/solvers/ode.html?highlight=constantsimp#sympy.solvers.ode.ode.constant_renumber) very compelling because they don't come from `dsolve` results (they just start by creating expressions involving `C1`, `C2`, `C3`) despite `constantsimp` being "written specifically to work with [dsolve()](https://docs.sympy.org/dev/modules/solvers/ode.html?highlight=constantsimp#sympy.solvers.ode.dsolve)". I understand the statement "There is no need to specify the dependent function, such as f(x), because it already has the independent symbol, x, in it." but think the examples would follow more naturally if they involved functions as `dsolve` typically returns such as $\cos(x)$.

For `constantsimp`, the best I came up with was adding one solution to a multiple of itself. Is there a more realistic situation where arbitrary constants could be simplified?

For `constant_renumber`, is there some example where applying `constantsimp` leads to removal of an arbitrary constant?

Reviewer: I don't think we should discuss either of these functions. They can be useful for end users in some situations, like if you are solving an ODE that comes from another ODE, so that you end up with constants from both solutions, but those situations are not typical. 

`constantsimp` and `constant_renumber` are already applied automatically to the result of `dsolve` (unless you specify `simplify=False`; and unlike `solve`, there's rarely a reason to need to do that).

# Solve an Ordinary Differential Equation (ODE) Algebraically

Use SymPy to solve an ordinary differential equation (ODE) algebraically. For
example, solving $y''(x) + 9y(x)=0 $ yields $ y(x)=C_{1} \sin(3x)+ C_{2}
\cos(3x)$.

## Alternatives to Consider
- To numerically solve a system of ODEs, use a [SciPy ODE
  solver](https://docs.scipy.org/doc/scipy/reference/integrate.html#solving-initial-value-problems-for-ode-systems)

>   such as `solve_ivp`.

Reviewer 3: A super common thing to do is to create an ode in sympy, lambdify it for use in solve_ivp, and then integrate numerically. I think it could be helpful for us to show that here (or another guide). We have some examples in the code generation tutorial: https://www.sympy.org/scipy-2017-codegen-tutorial/ and here is an example from my course materials: https://moorepants.github.io/learn-multibody-dynamics/simulation.html.

Author: Thanks, @moorepants. I would be happy to include such an example on this page. I'm having a tough time converting an [example from the code generation tutorial](https://www.sympy.org/scipy-2017-codegen-tutorial/notebooks/25-chemical-kinetics-intro.html#Solving-the-initial-value-problem-numerically) from [odeint](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.odeint.html?highlight=odeint#scipy.integrate.odeint) (SciPy's outdated function) to [solve_ivp](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.solve_ivp.html#scipy.integrate.solve_ivp) and using [lambdify](https://docs.sympy.org/latest/modules/utilities/lambdify.html?highlight=lambdify#module-sympy.utilities.lambdify). Here's what I've tried and the error it gives. Any assistance would be greatly appreciated.

```py
import sympy as sym
sym.init_printing()

def rhs(y, t, kf, kb):
    rf = kf * y[0]**2 * y[1]
    rb = kb * y[2]**2
    return [2*(rb - rf), rb - rf, 2*(rf - rb)]

y, k = sym.symbols('y:3'), sym.symbols('kf kb')
ydot = rhs(y, None, *k)

t = sym.symbols('t')  # not used in this case.
f = sym.lambdify((y, t) + k, ydot)  # EXERCISE: (y, t) + k

import numpy as np
from scipy.integrate import solve_ivp
tout = np.linspace(0, 10)
k_vals = np.array([0.42, 0.17])  # arbitrary in this case
y0 = [1, 1, 0]

solve_ivp(f, tout, y0, args=k_vals)
```

```py
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/var/folders/9w/cn8tqbv976gckxd0c9cbr1sw0000gn/T/ipykernel_96785/228934301.py in <module>
----> 1 solve_ivp(f, tout, y0, args=k_vals)

~/opt/anaconda3/lib/python3.9/site-packages/scipy/integrate/_ivp/ivp.py in solve_ivp(fun, t_span, y0, method, t_eval, dense_output, events, vectorized, args, **options)
    540         method = METHODS[method]
    541 
--> 542     solver = method(fun, t0, y0, tf, vectorized=vectorized, **options)
    543 
    544     if t_eval is None:

~/opt/anaconda3/lib/python3.9/site-packages/scipy/integrate/_ivp/rk.py in __init__(self, fun, t0, y0, t_bound, max_step, rtol, atol, vectorized, first_step, **extraneous)
     92         self.max_step = validate_max_step(max_step)
     93         self.rtol, self.atol = validate_tol(rtol, atol, self.n)
---> 94         self.f = self.fun(self.t, self.y)
     95         if first_step is None:
     96             self.h_abs = select_initial_step(

~/opt/anaconda3/lib/python3.9/site-packages/scipy/integrate/_ivp/base.py in fun(t, y)
    136         def fun(t, y):
    137             self.nfev += 1
--> 138             return self.fun_single(t, y)
    139 
    140         self.fun = fun

~/opt/anaconda3/lib/python3.9/site-packages/scipy/integrate/_ivp/base.py in fun_wrapped(t, y)
     18 
     19     def fun_wrapped(t, y):
---> 20         return np.asarray(fun(t, y), dtype=dtype)
     21 
     22     return fun_wrapped, y0

~/opt/anaconda3/lib/python3.9/site-packages/scipy/integrate/_ivp/ivp.py in <lambda>(t, x, fun)
    512         # additional parameters.  Pass in the original fun as a keyword
    513         # argument to keep it in the scope of the lambda.
--> 514         fun = lambda t, x, fun=fun: fun(t, x, *args)
    515         jac = options.get('jac')
    516         if callable(jac):

<lambdifygenerated-4> in _lambdifygenerated(_Dummy_26, t, kf, kb)
      1 def _lambdifygenerated(_Dummy_26, t, kf, kb):
----> 2     [y0, y1, y2] = _Dummy_26
      3     return [2*kb*y2**2 - 2*kf*y0**2*y1, kb*y2**2 - kf*y0**2*y1, -2*kb*y2**2 + 2*kf*y0**2*y1]

TypeError: cannot unpack non-iterable float object
```

Reviewer 3: This runs (but double check I didn't change the math).

solve_ivp takes t, y not y, t as args and the time is given as a tuple of (t0, tf) not an array. They changed several things about the api compared to odeint, which is a bit confusing.

```python
import sympy as sym
sym.init_printing()

def rhs(t, y, kf, kb):
    rf = kf * y[0]**2 * y[1]
    rb = kb * y[2]**2
    return [2*(rb - rf), rb - rf, 2*(rf - rb)]

y, k = sym.symbols('y:3'), sym.symbols('kf kb')
ydot = rhs(None, y, *k)

t = sym.symbols('t')  # not used in this case.
f = sym.lambdify((t, y) + k, ydot)  # EXERCISE: (y, t) + k

import numpy as np
from scipy.integrate import solve_ivp
k_vals = np.array([0.42, 0.17])  # arbitrary in this case
y0 = [1, 1, 0]

solve_ivp(f, (0, 10), y0, args=k_vals)
```

Reviewer 3: This removes the k values unpacking, maybe a bit clearer for learners.

```python
import sympy as sym
sym.init_printing()

def rhs(t, y, kf, kb):
    rf = kf * y[0]**2 * y[1]
    rb = kb * y[2]**2
    return [2*(rb - rf), rb - rf, 2*(rf - rb)]

y, (kf, kb) = sym.symbols('y:3'), sym.symbols('kf kb')
ydot = rhs(None, y, kf, kb)  # show that ydot is symbolic here

t = sym.symbols('t')  # not used in this case.
f = sym.lambdify((t, y, kf, kb), ydot)

import numpy as np
from scipy.integrate import solve_ivp
k_vals = np.array([0.42, 0.17])  # arbitrary in this case
y0 = [1, 1, 0]

solve_ivp(f, (0, 10), y0, args=k_vals)
```

Author: Great, thanks!

     6.37253319e-01 1.27438822e+00 2.15637690e+00 3.28555351e+00
     4.69240977e+00 6.45455786e+00 8.73068099e+00 1.00000000e+01], 't_events': None, 'y': [[1.         0.99858969 0.98475531 0.86889862 0.68120238 0.55390608
      0.47951256 0.44569558 0.43354565 0.43020361 0.42955182 0.4294468 ]
     [1.         0.99929485 0.99237766 0.93444931 0.84060119 0.77695304
      0.73975628 0.72284779 0.71677282 0.71510181 0.71477591 0.7147234 ]
     [0.         0.00141031 0.01524469 0.13110138 0.31879762 0.44609392
      0.52048744 0.55430442 0.56645435 0.56979639 0.57044818 0.5705532 ]], 'y_events': None}
```

`ydot` is the derivative of the function `y`, and the value of `ydot` is given
by the function `rhs` (right-hand side) for input values `y`, `kf`, and `kb`. We
use {func}`~.lambdify` to convert the SymPy symbolic expression for `ydot` into
a form that SciPy can evaluate numerically, `f`. Finally, we call SciPy's
`solve_ivp` by passing it the function `f`, the interval of integration, the

> initial state, and the arguments to pass to the function `f`.

Author: @moorepants I thought this code example merited some explanation. Could you please check that what I wrote explains it well?

[{y(x): C1 - C2*exp(x), z(x): C2*exp(x)}, {y(x): C1 + C2*exp(x), z(x): C2*exp(x)}]
>>> solutions_list = [{fn.lhs:fn.rhs for fn in solution} for solution in solutions]
>>> solutions_list # nested comprehension approach
[{y(x): C1 - C2*exp(x), z(x): C2*exp(x)}, {y(x): C1 + C2*exp(x), z(x): C2*exp(x)}]
>>> solutions_list[0][y(x)]
C1 - C2*exp(x)
```

### Work With Arbitrary Constants

You can manipulate arbitrary constants such as `C1`, `C2`, and `C3`, which are
generated automatically by {func}`~.dsolve`, by creating them as symbols. For
example, if you want to assign values to arbitrary constants, you can create
them as symbols and then substitute in their values using
{meth}`~sympy.core.basic.Basic.subs`:

```py
>>> from sympy import Function, dsolve, Derivative, symbols, pi
>>> x, C1, C2 = symbols("x, C1, C2")
>>> y = Function('y')
>>> result = dsolve(Derivative(y(x), x, x) + 9*y(x), y(x)).rhs
>>> result
C1*sin(3*x) + C2*cos(3*x)
>>> result.subs({C1: 7, C2: pi})
7*sin(3*x) + pi*cos(3*x)
```

### Numerically Solve an ODE in SciPy

A common workflow which leverages
[SciPy's](https://docs.scipy.org/doc/scipy/index.html) fast numerical ODE
solving is
1. set up an ODE in SymPy

> 2. convert it to a lambda function using {func}`~.lambdify`

Reviewer 3: ```suggestion
2. convert it to a numerical function using {func}`~.lambdify`
```

> 3. solve it numerically using SciPy's `solve_ivp`.

Reviewer 3: ```suggestion
3. solve the initial value problem by numerically integrating the ODE using SciPy's `solve_ivp`.
```

```{warning}
{func}`~.lambdify` uses {external:func}`~.exec` to dynamically execute Python code, and thus should not be used on unsanitized input.
```

Here is an [example from the field of chemical

> kinetics](https://www.sympy.org/scipy-2017-codegen-tutorial/notebooks/25-chemical-kinetics-intro.html):

Reviewer 3: ```suggestion
kinetics](https://www.sympy.org/scipy-2017-codegen-tutorial/notebooks/25-chemical-kinetics-intro.html) 
where the nonlinear ordinary differential equations take this form:

$$
r_f = & k_f y_0(t)^2  y_1(t) \\
r_b = & k_b y_2(t)^2 \\
\frac{d y_0(t)}{dt} = & 2(r_b - r_f) \\
\frac{d y_1(t)}{dt} = & r_b - r_f \\
\frac{d y_2(t)}{dt} = & 2(r_f - r_b)
$$

and

$$

\vec{y}(t) = \begin{bmatrix}
y_0(t) \\
y_1(t) \\
y_2(t)
\end{bmatrix}
$$
```

```py
>>> from sympy import symbols, lambdify
>>> import numpy as np
>>> import scipy.integrate
>>> def rhs(t, y, kf, kb):
...     rf = kf * y[0]**2 * y[1]
...     rb = kb * y[2]**2
...     return [2*(rb - rf), rb - rf, 2*(rf - rb)]
>>> y, (kf, kb) = symbols('y:3'), symbols('kf kb')

> >>> ydot = rhs(None, y, kf, kb)

Reviewer 3: ```suggestion
>>> ydot = rhs(None, y, kf, kb)
>>> ydot
```

Display ydot so that they see it is symbolic at this stage.

Reviewer 3: Also, passing in `None` for `t` is probably not clear for a learner. Using the same `rhs` to both construct the symbolic form and evaluate the numerics is a bit of a trick that will be lost on learners. I think creating the symbolic form separately and explicitly, as you do in prior examples, is more clear for a learner. If you do that, then you can avoid these tricks for shortening the code.

>>> t = symbols('t') # not used in this case
>>> f = lambdify((t, y, kf, kb), ydot)
>>> k_vals = np.array([0.42, 0.17]) # arbitrary in this case

> >>> y0 = [1, 1, 0]

Reviewer 3: call this the initial condition somewhere

>>> scipy.integrate.solve_ivp(f, (0, 10), y0, args=k_vals)
    {'message': 'The solver successfully reached the end of the integration interval.', 'nfev': 68, 'njev': 0, 'nlu': 0, 'sol': None, 'status': 0, 'success': True, 't': [0.00000000e+00 1.68190462e-03 1.85009508e-02 1.86691413e-01
     6.37253319e-01 1.27438822e+00 2.15637690e+00 3.28555351e+00
     4.69240977e+00 6.45455786e+00 8.73068099e+00 1.00000000e+01], 't_events': None, 'y': [[1.         0.99858969 0.98475531 0.86889862 0.68120238 0.55390608
      0.47951256 0.44569558 0.43354565 0.43020361 0.42955182 0.4294468 ]
     [1.         0.99929485 0.99237766 0.93444931 0.84060119 0.77695304
      0.73975628 0.72284779 0.71677282 0.71510181 0.71477591 0.7147234 ]
     [0.         0.00141031 0.01524469 0.13110138 0.31879762 0.44609392
      0.52048744 0.55430442 0.56645435 0.56979639 0.57044818 0.5705532 ]], 'y_events': None}

> ```

Reviewer 3: We have the matplolib extension installed so you can create graphs. This data is more easily understood if you make a simple plot. See the beam module docs for examples of using the plotting extension (or the plotting module).

Author: Thanks. I made a graph which displays properly in a Jupyter Notebook. But I'm having trouble getting the plot to display properly via Sphinx. I was able to get a simple matplotlib example to display, but not the ODE solution. Also, may not be relevant once the ODE plot works, but if I make a second simple plot with different values, it shows the first plot again:

<img width="530" alt="matplotlib for ODE" src="https://user-images.githubusercontent.com/1335522/191617582-3707e842-1467-4769-b4b3-44bc862ffd6d.png">

Reviewer 3: There are flags to control whether figures are closed. We got it working in the beam pages.

Author: One problem is that any plot code block containing scipy fails due to:

`ModuleNotFoundError: No module named 'scipy'`

(A simple plot block like `plt.plot([11, 12, 13], [16, 15, 14])` renders a graph image fine if I comment out the scipy import; numpy and matplotlib imports work fine.) Is there some way to make a package available inside a plot block? Perhaps this [configuration option](https://matplotlib.org/3.1.3/devel/plot_directive.html#configuration-options):

> plot_working_directory
>     By default, the working directory will be changed to the directory of the example, so the code can get at its data files, if any. Also its path will be added to [sys.path](https://docs.python.org/3/library/sys.html#sys.path) so it can import any helper modules sitting beside it. This configuration option can be used to specify a central directory (also added to [sys.path](https://docs.python.org/3/library/sys.html#sys.path)) where data files and helper modules for all code are located.

Author: > There are flags to control whether figures are closed. We got it working in the beam pages.

Thanks--yes, by adding `:context: close-figs` to each plot directive as in the [beam module](https://github.com/sympy/sympy/blob/26f7bdbe3f860e7b4492e102edec2d6b429b5aaf/sympy/physics/continuum_mechanics/beam.py#L1314), different figures can be created. Probably not needed for this page because I plan on including only one plot; good to know how to do.

Author: Turned out that SciPy needed to be a dependency for the docs, by adding `scipy` to `doc/requirements.txt`.

`ydot` is the derivative of the function `y`, and the value of `ydot` is given
by the function `rhs` (right-hand side) for input values `y`, `kf`, and `kb`. We
use {func}`~.lambdify` to convert the SymPy symbolic expression for `ydot` into
a form that SciPy can evaluate numerically, `f`. Finally, we call SciPy's
`solve_ivp` by passing it the function `f`, the interval of integration, the
initial state, and the arguments to pass to the function `f`. SciPy's
`solve_ivp` returns a result containing `t` (time) points and corresponding `y`

> (numerical function result) values for each initial point in `y0`.

Author: I believe I'm interpreting the [SciPy solve_ivp](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.solve_ivp.html#scipy.integrate.solve_ivp) documentation correctly that each sub-list in `y` corresponds to an initial point in `y0`. Then why are the results different for the first two sub-lists when each corresponds to a `y0` value of `1`?

Reviewer 3: I would interweave these explanations with each line of the code, i.e. state what each line is doing as you do it.

(solving-guide-inequalities)=
# Solve an Ordinary Differential Equation (ODE) Algebraically

Use SymPy to solve an ordinary differential equation (ODE) algebraically. For
example, solving $y''(x) + 9y(x)=0 $ yields $ y(x)=C_{1} \sin(3x)+ C_{2}
\cos(3x)$.

## Alternatives to Consider
- To numerically solve a system of ODEs, use a [SciPy ODE
  solver](https://docs.scipy.org/doc/scipy/reference/integrate.html#solving-initial-value-problems-for-ode-systems)
  such as `solve_ivp`. You can also use SymPy to create and then
  {func}`~.lambdify` an ODE to be solved numerically using SciPy's as
  `solve_ivp` as described below in [](#numerically-solve-an-ode-in-scipy).

## Solve an Ordinary Differential Equation (ODE)

> Here is an example of solving an ordinary differential equation algebraically

Reviewer: ```suggestion
Here is an example of solving the above ordinary differential equation algebraically
```

using {func}`~.dsolve`:

```py
>>> from sympy import Function, dsolve, Derivative
>>> from sympy.abc import x
>>> y = Function('y')
>>> result = dsolve(Derivative(y(x), x, x) + 9*y(x), y(x))
>>> result
Eq(y(x), C1*sin(3*x) + C2*cos(3*x))
```

You can then use SymPy to verify that the solution is correct:

```py
>>> from sympy import checkodesol

> >>> solution = result.rhs

Reviewer: It's not necessary to do this. `checkodesol` can accept the direct Eq output of `dsolve`.

>>> solution
C1*sin(3*x) + C2*cos(3*x)
>>> checkodesol(Derivative(y(x), x, x) + 9*y(x), solution)
(True, 0)
```

The output of {func}`~.checkodesol` is a tuple where the first item, a boolean,
tells whether substituting the solution into the ODE results in `0`, indicating
the solution is correct.

> ## Guidance

Reviewer: I'm a little unclear how the "Guidance" section is supposed to work on these pages. How come on this page the subsequent sections are on the same level as "Guidance", but on the "solving systems of equations numerically" page several sections are under "Guidance"?

Author: The intent of the "Guidance" section is to have zero or more subsections that provide tips on how to set up using the function in question. 

- On this ODE page, I identified only one tip, so there's just one subsection, "Input Format"; so "Guidance" actually has one child section. The subsequent sections are not (in my way of thinking) that kind of guidance, so they're at the same level (`h2`) as "Guidance".
- On the ["solving systems of equations numerically" page](https://docs.sympy.org/dev/guides/solving/solve-numerically.html), I identified several such tips, so they're all children sections of "Guidance".

The (`h2`) sections after "Guidance" are typically "Use the Solution Result", "Tradeoffs", etc. I admit it's a somewhat subjective decision what counts as Guidance. I used a template with sections and slotted content into what I thought was the most logical section. We can revisit if needed.

### Input Format

There are many ways to express derivatives of functions. For an undefined
function, both {class}`~.Derivative` and {func}`~.diff` represent the undefined
derivative. Thus, all of the following `ypp` ("y prime prime") represent $y''$,
the second derivative with respect to $x$ of a function $y(x)$:

```py
ypp = y(x).diff(x, x)
ypp = y(x).diff(x, 2)
ypp = y(x).diff((x, 2))
ypp = diff(y(x), x, x)
ypp = diff(y(x), x, 2)
ypp = Derivative(y(x), x, x)
ypp = Derivative(y(x), x, 2)

> ypp = Derivative(Derivative(y(y(x), x), x)

Reviewer: ```suggestion
ypp = Derivative(Derivative(y(x), x), x)
```

ypp = diff(diff(y(x), x), x)
yp = y(x).diff(x)
ypp = yp.diff(x)
```

We recommend specifying the function to be solved for, as the second argument to
{func}`~.dsolve`. Note that it must be a function rather than a variable
(symbol). SymPy will give an error if you specify a variable ($x$) rather than a
function ($f(x)$):

```py
>>> dsolve(Derivative(y(x), x, x) + 9*y(x), x)
Traceback (most recent call last):
    ...
ValueError: dsolve() and classify_ode() only work with functions of one variable, not x
```

Similarly, you must specify the argument of the function: $f(x)$, not just $f$.

You can define the function to be solved for in two ways. The subsequent syntax
for specifying initial conditions depends on your choice.

> ## Define a Function Without Including Its Independent Variable

Reviewer: I think this is supposed to be level 4 (and the same with the next several headers).

Author: My intention was to have

1. Define a Function Without Including Its Independent Variable
2. Define a Function of an Independent Variable

as level 2, with the idea that each is an _option_ (as you suggest below) for an entire approach to using dsolve--setting up `dsolve`, adding initial conditions or boundary conditions, and (for option 2) a caveat. The subsequent sections (Use the Solution Result, etc.) apply to either option 2.

If you think it would be helpful, I could create a parent level 2 and put the two options below it as level 3:
## Options to Define an ODE
### Option 1: Define a Function Without Including Its Independent Variable
### Option 2: Define a Function of an Independent Variable

Author: Move to a new h2 Options to Define an ODE
You can define the function to be solved for in two ways. The subsequent syntax for specifying initial conditions depends on your choice.

Could move options 1 and 2 under Guidance instead

Rename "Input Format" to Defining Derivatives

> As in the example above, you can define a function without including its

Reviewer: It's a little confusing what this means. It took me a moment to realize this is talking specifically about the line `f = Function('f')`. The independent variable (x) is used elsewhere, just not on that line.

independent variable:

```py
>>> from sympy import symbols, Eq, Function, dsolve
>>> f, g = symbols("f g", cls=Function)
>>> x = symbols("x")
>>> eqs = [Eq(f(x).diff(x), g(x)), Eq(g(x).diff(x), f(x))]
>>> dsolve(eqs, [f(x), g(x)])
[Eq(f(x), -C1*exp(-x) + C2*exp(x)), Eq(g(x), C1*exp(-x) + C2*exp(x))]
```

Note that you supply the functions to be solved for as a list as the second
argument of {func}`~.dsolve`, here `[f(x), g(x)]`.

> ### Specify Initial (Boundary) Conditions

Reviewer: ```suggestion
### Specify Initial Conditions or Boundary Conditions
```

> If your differential equation(s) have initial or boundary conditions, specify

Reviewer: It should be stressed here that both initial conditions and boundary conditions are treated in the same way (even though the argument is called `ics`).

them with the {func}`~.dsolve` optional argument `ics`. It should be given in
the form of `{f(x0): y0, f(x).diff(x).subs(x, x1): y1}` and so on where, for

> example, the value of `f(x)` at $x = $ `x0` is `y0`. For power series solutions,

Reviewer: Don't mix LaTeX and code like this. Use either one or the other.

Reviewer: Should also note here that `f'(x1) = y1` (that's the one that is less clear).

Author: My interpretation was that

- `code` is for code, e.g. `dsolve`, as in the [documentation style guide](https://docs.sympy.org/dev/contributing/documentation-style-guide.html#code)
- LaTeX is for math, e.g. $\pi$, similar to the [docstring style guide](https://docs.sympy.org/dev/contributing/docstring.html#docstrings-for-classes-that-are-mathematical-functions)

Does that make sense?

Author: For e.g. x = x0, make 0 a subscript. Use math.

if no initial conditions are specified `f(0)` is assumed to be `C0` and the
power series solution is calculated about $0$.

Here is an example of setting the initial values for functions, namely namely
$f(0) = 1$ and $g(2) = 3$:

```py
>>> from sympy import symbols, Eq, Function, dsolve
>>> f, g = symbols("f g", cls=Function)
>>> x = symbols("x")
>>> eqs = [Eq(f(x).diff(x), g(x)), Eq(g(x).diff(x), f(x))]
>>> dsolve(eqs, [f(x), g(x)])
[Eq(f(x), -C1*exp(-x) + C2*exp(x)), Eq(g(x), C1*exp(-x) + C2*exp(x))]
>>> dsolve(eqs, [f(x), g(x)], ics={f(0): 1, g(2): 3})
[Eq(f(x), (1 + 3*exp(2))*exp(x)/(1 + exp(4)) - (-exp(4) + 3*exp(2))*exp(-x)/(1 + exp(4))), Eq(g(x), (1 + 3*exp(2))*exp(x)/(1 + exp(4)) + (-exp(4) + 3*exp(2))*exp(-x)/(1 + exp(4)))]
```

Here is an example of setting the initial value for the derivative of a

> function, namely $f'(1) = 2$:

Reviewer: Even though it's supported, I don't think a typical use-case would be to underspecify the initial conditions like this. I would either also add a second condition or change this example to use a single ODE.

Author: Either specify an initial condition for g'(x), or just use one equation.

variable (for example $t$), so that `x`, represents `x(t)`:

```py
>>> from sympy import symbols, Function, dsolve
>>> t = symbols('t')
>>> x = Function('x')(t)
>>> x
x(t)
>>> xp = x.diff()
>>> xpp = xp.diff()
>>> eq = xpp + 2*xp + x
>>> eq
x(t) + 2*Derivative(x(t), t) + Derivative(x(t), (t, 2))
>>> dsolve(eq, x)
Eq(x(t), (C1 + C2*t)*exp(-t))
```

Using this convention, the second argument of {func}`~.dsolve`, `x`, represents
`x(t)`, so SymPy recognizes it as a valid function to solve for.

### Specify Initial (Boundary) Conditions

Using that syntax, you specify initial conditions by substituting in values of
the independent variable using {func}`~sympy.core.basic.Basic.subs` because the
function $x$ already has its independent variable as an argument $t$:

```py
>>> dsolve(eq, x, ics={x.subs(t, 0): 0})
Eq(x(t), C2*t*exp(-t))
```

### Beware Copying and Pasting Results

If you choose to define a function of an independent variable, note that copying
a result and pasting it into subsequent code may cause an error because `x` is
already defined as `x(t)`, so if you paste in `x(t)` it is interpreted as
`x(t)(t)`:

```py

> >>> dsolve(x(t).diff(), x)

Reviewer: The no-argument variant of `diff()` that doesn't specify the symbol to differentiate wrt is provided as a convenience, but it's generally considered bad practice, so we should avoid using it in documentation.

Traceback (most recent call last):
    ...
TypeError: 'x' object is not callable
```

So remember to exclude the independent variable call `(t)`:

```py
>>> dsolve(x.diff(), x)
Eq(x(t), C1)
```

## Use the Solution Result

Unlike other solving functions, {func}`~.dsolve` returns an {class}`~.Equality`
(equation) formatted as, for example, `Eq(y(x), C1*sin(3*x) + C2*cos(3*x))`
which is equivalent to the mathematical notation $y(x) = C_1 \sin(3x) + C_2
\cos(3x)$.

### Extract the Result for One Solution and Function

You can extract the result from an {class}`~.Equality` using the right-hand side

> property {any}`rhs <sympy.core.relational.Relational.rhs>`:

Reviewer: I don't know if we should caution here that the result may not actually be solved for f(x). For cases when SymPy cannot solve the equation explicitly, the solution will be returned in implicit form.

Author: Plan: Add a section and example where SymPy can't find explicit solution--just an implicit solution. Just find a function that can't be solved, make into a differential equation for y; for example, x - cos(x) which can't be solved algebraically.

Reviewer: Here's an example 

```
>>> dsolve(f(x).diff(x) + exp(-f(x))*f(x))
Eq(Ei(f(x)), C1 - x)
```

Here Ei is the exponential integral function, which doesn't have a closed-form inverse as far as I know. It looks like dsolve doesn't do anything in this case. It just returns a solution that's something equals something, where either side could contain f(x) (i.e., an [implicit solution](https://en.wikipedia.org/wiki/Implicit_function)).

```py
>>> from sympy import Function, dsolve, Derivative
>>> from sympy.abc import x
>>> y = Function('y')
>>> result = dsolve(Derivative(y(x), x, x) + 9*y(x), y(x))
>>> result
Eq(y(x), C1*sin(3*x) + C2*cos(3*x))
>>> result.rhs
C1*sin(3*x) + C2*cos(3*x)
```

### Extract the Result for Multiple Function-Solution Pairs

If you are solving a system of equations with multiple unknown functions,

> {func}`~.dsolve` will return a nested list of equalities, the outer list

Reviewer: It only returns a nested list when there are multiple different solution sets. When there is only one solution set it just returns a single list (for example, if you replace $y'(x)^2 = z(x)^2$ with $y'(x) = z(x)^2$  in this example).

  >>> t_eval = np.linspace(0, 10, 50) # evaluate integral from t = 0-10 for 50 points
  >>> # Call SciPy's ODE initial value problem solver solve_ivp by passing it
  >>> #   the function f,
  >>> #   the interval of integration,
  >>> #   the initial state, and
  >>> #   the arguments to pass to the function f
  >>> solution = scipy.integrate.solve_ivp(f, (0, 10), y0, t_eval=t_eval, args=k_vals)
  >>> # Extract the y (concentration) values from SciPy solution result
  >>> y = solution.y
  >>> # Plot the result graphically using matplotlib
  >>> plt.plot(t_eval, y.T) # doctest: +SKIP
  >>> # Add title, legend, and axis labels to the plot
  >>> plt.title('Chemical Kinetics') # doctest: +SKIP
  >>> plt.legend(['NO', 'Br$_2$', 'NOBr'], shadow=True) # doctest: +SKIP
  >>> plt.xlabel('time') # doctest: +SKIP
  >>> plt.ylabel('concentration') # doctest: +SKIP
  >>> # Finally, display the annotated plot
  >>> plt.show()
  ```

SciPy's `solve_ivp` returns a result containing `y` (numerical function result,
here, concentration) values for each of the three chemical species,
corresponding to the time points `t_eval`.

## Ordinary Differential Equation Solving Hints

> ### Return Unevaluated Integrals

Reviewer: We really just need to add this as a flag to dsolve so you don't have to mess with hints to do this.

Author: That would be nice.

```

Using this convention, the second argument of {func}`~.dsolve`, `x`, represents
`x(t)`, so SymPy recognizes it as a valid function to solve for.

### Specify Initial (Boundary) Conditions

Using that syntax, you specify initial conditions by substituting in values of
the independent variable using {func}`~sympy.core.basic.Basic.subs` because the
function $x$ already has its independent variable as an argument $t$:

```py
>>> dsolve(eq, x, ics={x.subs(t, 0): 0})
Eq(x(t), C2*t*exp(-t))
```

### Beware Copying and Pasting Results

If you choose to define a function of an independent variable, note that copying
a result and pasting it into subsequent code may cause an error because `x` is
already defined as `x(t)`, so if you paste in `x(t)` it is interpreted as
`x(t)(t)`:

```py
>>> dsolve(x(t).diff(), x)
Traceback (most recent call last):
    ...
TypeError: 'x' object is not callable
```

So remember to exclude the independent variable call `(t)`:

```py
>>> dsolve(x.diff(), x)
Eq(x(t), C1)
```

## Use the Solution Result

Unlike other solving functions, {func}`~.dsolve` returns an {class}`~.Equality`
(equation) formatted as, for example, `Eq(y(x), C1*sin(3*x) + C2*cos(3*x))`

> which is equivalent to the mathematical notation $y(x) = C1 \sin(3x) + C2

Reviewer: Make sure to subscript the constants in LaTeX (`C_1`). You can also generally just use `print(latex(expr))` on an expression to get a LaTeX form to put in the docs.

...     solution_dict = {}
...     for fn in solution:
...             solution_dict.update({fn.lhs: fn.rhs})
...     solutions_list.append(solution_dict)
>>> solutions_list
[{y(x): C1 - C2*exp(x), z(x): C2*exp(x)}, {y(x): C1 + C2*exp(x), z(x): C2*exp(x)}]
>>> solutions_list = [{fn.lhs:fn.rhs for fn in solution} for solution in solutions]
>>> solutions_list # nested comprehension approach
[{y(x): C1 - C2*exp(x), z(x): C2*exp(x)}, {y(x): C1 + C2*exp(x), z(x): C2*exp(x)}]
>>> solutions_list[0][y(x)]
C1 - C2*exp(x)
```

### Work With Arbitrary Constants

You can manipulate arbitrary constants such as `C1`, `C2`, and `C3`, which are
generated automatically by {func}`~.dsolve`, by creating them as symbols. For
example, if you want to assign values to arbitrary constants, you can create
them as symbols and then substitute in their values using
{meth}`~sympy.core.basic.Basic.subs`:

```py
>>> from sympy import Function, dsolve, Derivative, symbols, pi
>>> x, C1, C2 = symbols("x, C1, C2")
>>> y = Function('y')
>>> result = dsolve(Derivative(y(x), x, x) + 9*y(x), y(x)).rhs
>>> result
C1*sin(3*x) + C2*cos(3*x)
>>> result.subs({C1: 7, C2: pi})
7*sin(3*x) + pi*cos(3*x)
```

> ### Numerically Solve an ODE in SciPy

Reviewer: I think this should be a level 2 header. It's not related to "use the solution result".

A common workflow which leverages
[SciPy's](https://docs.scipy.org/doc/scipy/index.html) fast numerical ODE
solving is
1. set up an ODE in SymPy
2. convert it to a numerical function using {func}`~.lambdify`
3. solve the initial value problem by numerically integrating the ODE using
   SciPy's `solve_ivp`.

> ```{warning}

Reviewer: I don't know if this needs to be repeated here.

\frac{d y_1(t)}{dt} = & r_b - r_f \\
\frac{d y_2(t)}{dt} = & 2(r_f - r_b) $$

and

$$

\vec{y}(t) = \begin{bmatrix} y_0(t) \\
y_1(t) \\
y_2(t) \end{bmatrix} $$

```{eval-rst}
.. plot:: 
  :format: doctest 
  :include-source: True

  >>> from sympy import symbols, lambdify
  >>> import numpy as np
  >>> import scipy.integrate
  >>> import matplotlib.pyplot as plt
  >>> def rhs(t, y, kf, kb):
  ...     rf = kf * y[0]**2 * y[1]
  ...     rb = kb * y[2]**2
  ...     return [2*(rb - rf), rb - rf, 2*(rf - rb)]
  >>> y, (kf, kb) = symbols('y:3'), symbols('kf kb')
  >>> ydot = rhs(None, y, kf, kb) # derivative of the function y(t); value given by function rhs (right-hand side) for input values y, kf, and kb
  >>> ydot
  [2*kb*y2**2 - 2*kf*y0**2*y1, kb*y2**2 - kf*y0**2*y1, -2*kb*y2**2 + 2*kf*y0**2*y1]
  >>> t = symbols('t') # not used in this case
  >>> f = lambdify((t, y, kf, kb), ydot) # convert the SymPy symbolic expression for ydot into a form that SciPy can evaluate numerically, f
  >>> k_vals = np.array([0.42, 0.17]) # arbitrary in this case
  >>> y0 = [1, 1, 0] # initial condition (initial values)
  >>> t_eval = np.linspace(0, 10, 50) # evaluate integral from t = 0-10 for 50 points

>   >>> solution = scipy.integrate.solve_ivp(f, (0, 10), y0, t_eval=t_eval, args=k_vals) # call SciPy's solve_ivp by passing it the function f, the interval of integration, the initial state, and the arguments to pass to the function f

Reviewer: It's easier to read long comments like this in doctest blocks when they are on their own line before the code, since text in doctest blocks does not text wrap, so any line that is too long will require you to scroll right to read it.

Author: Great, thanks for the tip. I thought the doctest would complain if you inserted a line with just a comment; seems fine (at least running doctest locally) here.

  >>> solution
      {'message': 'The solver successfully reached the end of the integration interval.', 'nfev': 68, 'njev': 0, 'nlu': 0, 'sol': None, 'status': 0, 'success': True, 't': [ 0.          0.20408163  0.40816327  0.6122449   0.81632653  1.02040816

>         1.2244898   1.42857143  1.63265306  1.83673469  2.04081633  2.24489796

Reviewer: You can replace unimportant chunks of the output with `...`

Author: Thanks. I plan on taking that one step further: Assuming the plot will work, I don't think it's even necessary to include the solution output at all.

        0.72021849 0.71919675 0.71838061 0.71773223 0.71721025 0.71676983
        0.71638533 0.71606247 0.71579958 0.71559248 0.71543446 0.71531629
        0.71522621 0.71514997 0.71507228 0.71499749 0.71493178 0.71487752
        0.71483596 0.71480726 0.71479042 0.71478338 0.71478292 0.71478472
        0.71478335 0.71477272 0.71475979 0.7147492  0.71474061 0.71473367
        0.71472805 0.7147234 ]
      [0.         0.1411542  0.23962891 0.31134362 0.36551142 0.40701812
        0.43929284 0.46487606 0.48521378 0.50136636 0.51428406 0.52479115
        0.5333617  0.54028258 0.54583936 0.55031198 0.55397486 0.5570396
        0.55956302 0.5616065  0.56323878 0.56453555 0.56557951 0.56646033
        0.56722934 0.56787506 0.56840084 0.56881503 0.56913108 0.56936743
        0.56954757 0.56970005 0.56985544 0.57000502 0.57013644 0.57024496
        0.57032808 0.57038549 0.57041915 0.57043324 0.57043417 0.57043056
        0.57043329 0.57045455 0.57048041 0.5705016  0.57051879 0.57053265
        0.57054389 0.5705532 ]], 'y_events': None}
  >>> y = solution.y # Extract the y (concentration) values from SciPy solution result
  >>> plt.plot(t_eval, y.T) # doctest: +SKIP
```

```{eval-rst}

> .. plot:: 

Reviewer: It should be possible to do this directly with myst. The rst directives translate directly into markdown syntax. See https://myst-parser.readthedocs.io/en/v0.13.6/using/syntax.html#directives-a-block-level-extension-point

Author: Yeah, I tried, using that format:

````
```{plot}
---
format: doctest 
include-source: True
---
>>> plt.plot([1, 2, 3], [4, 5, 6]) # doctest: +SKIP
```
````

and a slightly different one:

````
```{plot}
:format: doctest 
:include-source: True

>>> plt.plot([1, 2, 3], [4, 5, 6]) # doctest: +SKIP
```
````

and neither gave any output.

I asked on [Write the Docs (Slack for technical writing)](https://www.writethedocs.org/slack/) and the Executable Books lead developer @chrisjsewell replied:

> Yeh so from here: https://github.com/matplotlib/matplotlib/blob/269c0b94b4fcf8b1135011c1556eac29dc09de15/lib/matplotlib/sphinxext/plot_directive.py#L359
> the default `TEMPLATE` is hard-coded to be in rST.
> It looks like you can override this with the `plot_template` configuration option,
> or obviously you can raise a feature request with them to include such a template :sweat_smile:

Which I believe means someone would have to create a `TEMPLATE` which works for Markdown.

Reviewer: I actually remember having an issue with plotting in my custom functions guide too. That's good to know. So for now if we want to include plots here, we need to rewrite this page to use RST?

Author: No, thankfully it's fine for the page to be in Markdown as long as the plot directive is in RST--[a simple plot block renders a graph image fine](https://output.circle-artifacts.com/output/job/03cc4694-a200-4587-aa2b-bdd4fe6dd463/artifacts/0/doc/_build/html/guides/solving/solve-ode.html#numerically-solve-an-ode-in-scipy) using the wrapper:

````
```{eval-rst}
```
````

I think the remaining critical problem is importing scipy within the plot directive.
